src/include/serial_inline.h

```diff
@@ -36,10 +36,12 @@ __insert_simple_func(
          * against the next pointer might indicate that the skip list location is still valid, but
          * that may no longer be true when the atomic_cas operation executes.
          *
-         * Place a read barrier here to avoid this issue.
+         * Place a compiler barrier here to avoid this issue.
          */
-        WT_ORDERED_READ(old_ins, *ins_stack[i]);
-        if (old_ins != new_ins->next[i] || !__wt_atomic_cas_ptr(ins_stack[i], old_ins, new_ins))
+        old_ins = *ins_stack[i];
+        WT_C11_BARRIER();
+        if (old_ins != new_ins->next[i] ||
+          !__wt_atomic_memorder_cas_ptr(ins_stack[i], old_ins, new_ins, WT_ATOMIC_SEQ_CST))
```

Comment:
Does this really need SEQ_CST? My best guess is that a RELEASE is sufficient, but again I don't fully understand the context.

At the very least, I'm pretty sure that you would be fine with RELAXED on the failure case, especially since you don't use the actual value discovered by the swap.

Reply:
I think we at least we need acquire release? We need to synchronise the read with another write as well?

I think we should not allow any read or write being reordered across this.

Reply:
> I think we at least we need acquire release?  We need to synchronise the read with another write as well?

Is the write you are concerned with the next pass of the loop? If so, that would be handled by *its* release store happening after everything that occurred earlier on the same thread (including this release store). Is there a different write you are concerned with?

Also, I think we already have an acquire edge for this location/value pair if I understand this code correctly. At some point earlier on this thread did we do a load acquire on that address to get the value that we wrote into `new_ins->next[i]`? If so, then this CAS is already guaranteed to happen after the write of that value. And because of the CAS we know that the value must not have changed since then, so we don't need to re-establish the acquire edge to that address. I am assuming that we don't have to worry about the ABA problem here because we prevent freeing theses addresses while we are in there (I think that is handled by hazard pointers to the whole page).

> I think we should not allow any read or write being reordered across this.

Remember that the C model doesn't allow you to express exactly that. You need to think in terms of paired acquires and releases.

Reply:
I think you are right in the new c11 model context. However, the old wt memory model requires cas to give a full barrier semantic. This is why I'm a bit hesitant to relax the order here. As experiment, I'll have a try.

Reply:
> However, the old wt memory model requires cas to give a full barrier semantic.

So there are two forms of "requirements" and I think it is important to separate them:
1. The memory model has a definition that says that nothing at all can be reordered across a CAS. This is a requirement placed on the platform-specific implementations of the atomic primitives. And it is something that all algorithms can rely on whether they need it or not.
2. Each algorithm has its own specific requirements of what it needs from each operation. When translating the abstract algorithm to code in a specific memory model, you need to choose which operations to use to ensure that those operations meet the requirements of the algorithm. Of course, that is simpler if all ops have full barrier, totally-ordered semantics.

I think you are using "requires" in the sense of 1 here. But since the whole point is that we are transitioning between memory models, I think we really need to be thinking about 2. Rather than saying that the existing code has a full barrier and we must match it, I think it would be better to figure out what the algorithm really requires, and use those specific orderings. Of course it is possible that `seq_cst` really _is_ what the algorithm needs, but I think in many more cases we are getting it accidentally because it was the only option available in the memory model. And we can still choose to use `seq_cst` in most places out of an abundance of caution, but I think the work to reason through what the least-restrictive ordering is for all accesses (possibly excluding debug-only checks) will pay off.